### PR TITLE
feat: call coordinator service api

### PR DESCRIPTION
**Pull request by kittybest**
**Description**
try to call coordinator service api to make sure everything works well.

- [x] add and read coordinator service url in the env file
- [x] add function call to fetch public key of coordinator service
- [x] add function call to generate proof and succeed
- [x] merge function call

**Blocker**
PR #170 
https://github.com/privacy-scaling-explorations/maci/pull/1584

**Problem**
Deployed coordinator service in the local side, but always got `400 Bad Request`.
For the nest side, the error message is 
```
[Nest] 71842  - 2024/06/19 上午2:33:03   ERROR [AppController] Error:
[Nest] 71842  - 2024/06/19 上午2:33:03   ERROR [AppController] Error: contract runner does not support calling (operation="call", code=UNSUPPORTED_OPERATION, version=6.13.0)
```

**Related Issues**
#139

### CODE REVIEW

File: src/services/maci/service.ts [hunk 1]
```diff
@@ -684,6 +690,111 @@ export class MaciService {
...
+
+    const pollId = Number(id);
+
+    const network = await this.getNetwork();
+    const tallyContractAddress = this.storage.mustGetAddress(
+      EContracts.Tally,
+      network,
+      `poll-${pollId}`,
+    );
+    console.log("isMerged:", isMerged, "id:", pollId);
```
Comment (ctrlc03) [inline src/services/maci/service.ts:715]: should we remove console logs?

File: src/services/maci/service.ts [hunk 2]
```diff
@@ -684,6 +690,111 @@ export class MaciService {
...
+    if (!isMerged) {
+      // merge signups
+      try {
+        await mergeSignups({
+          pollId,
+          maciAddress: maciData.address,
+          signer: this.deployer,
+        });
+      } catch (error) {
+        console.log("Signups are already merged.", error);
```
Comment (ctrlc03) [inline src/services/maci/service.ts:726]: What if the error is something else? do we wanna start adding toasts for errors btw?

File: src/services/maci/service.ts [hunk 3]
```diff
@@ -684,6 +690,111 @@ export class MaciService {
...
+
+      // merge messages
+      try {
+        await mergeMessages({
+          pollId,
+          maciAddress: maciData.address,
+          signer: this.deployer,
+        });
+      } catch (error) {
+        console.log("Messages are already merged.", error);
```
Comment (ctrlc03) [inline src/services/maci/service.ts:737]: same as above

File: src/services/maci/service.ts [hunk 4]
```diff
@@ -684,6 +690,111 @@ export class MaciService {
...
+      }
+    }
+
+    // URL
+    const getPubKeyURL = `${config.coordinatorService}/v1/proof/publicKey`;
```
Comment (ctrlc03) [inline src/services/maci/service.ts:742]: should we move constants to a separate file?

**Review by ctrlc03 — COMMENTED** [review]
Thanks, left some comments

File: src/services/maci/service.ts (hunk 2 above)
Comment (kittybest) [reply src/services/maci/service.ts:726]: I think we should just log the error now.

File: package.json [hunk 5]
```diff
@@ -36,13 +36,14 @@
     "@vercel/blob": "^0.19.0",
     "clsx": "^2.1.0",
     "cmdk": "^0.2.0",
+    "crypto": "^1.0.1",
```
[the remark this answers is not in the document]
Comment (kittybest) [reply package.json:39]: removed

**Comment by kittybest** [comment]
image

⬆️ this is the current problem, already made a PR for it: https://github.com/privacy-scaling-explorations/maci/pull/1591

**Comment by kittybest** [comment]
**Remained Problem**
```
const { id, isMerged } = await getPoll({
      maciAddress: maciData.address,
      signer: this.deployer,
      provider: this.deployer.provider,
    });
```
the `isMerged` returns only the `pollContract.mergedStateRoot()`, should also check if the messages are also merged.

**Review by ctrlc03 — APPROVED** [review]
Thanks

File: src/services/maci/service.ts [hunk 6]
```diff
@@ -684,6 +691,98 @@ export class MaciService {
...
+    const ret = await fetch(GENPROOF_URL, {
+      method: "POST",
+      headers,
+      body: JSON.stringify(body),
+    }).then(async (r) => {
+      if (!r.ok) throw new Error("Gen proof error");
+      return r.json();
+    });
+
+    return ret;
```
Comment (0xmad) [inline src/services/maci/service.ts:783]:
```suggestion
    return fetch(GENPROOF_URL, {
      method: "POST",
      headers,
      body: JSON.stringify(body),
    }).then(async (r) => {
      if (!r.ok) throw new Error("Gen proof error");
      return r.json();
    });
```

File: src/services/maci/service.ts [hunk 7]
```diff
@@ -684,6 +691,96 @@ export class MaciService {
...
+    };
+
+    return (ret = await fetch(GENPROOF_URL, {
+      method: "POST",
+      headers,
+      body: JSON.stringify(body),
+    }).then(async (r) => {
+      if (!r.ok) throw new Error("Gen proof error");
+      return r.json();
+    }));
```
Comment (0xmad) [inline src/services/maci/service.ts:781]:
```suggestion
    return fetch(GENPROOF_URL, {
      method: "POST",
      headers,
      body: JSON.stringify(body),
    }).then(async (r) => {
      if (!r.ok) throw new Error("Gen proof error");
      return r.json();
    });
```

**Review by 0xmad — APPROVED** [review]
@kittybest thanks!

**Pull request merged**